### PR TITLE
Update Build Expression parser for forward compatibility

### PR DESCRIPTION
**Pull request by MDrakos**
<table>
<td>
  <a href="https://activestatef.atlassian.net/browse/DX-2059" title="DX-2059" target="_blank">StoryDX-2059</a>  Update Build Expression Parser to be forward compatible
  </td></table>
  <br />

### CODE REVIEW

**Review by mitchell-as — CHANGES_REQUESTED** [review]
Overall I like what you've done here. There are a few improvements I think you can make though.

File: pkg/platform/runtime/buildexpression/buildexpression.go [hunk 1]
```diff
@@ -41,10 +41,12 @@ const (
 var funcNodeNotFoundError = errors.New("Could not find function node")
 
 type BuildExpression struct {
-	Let *Let
+	Let    *Let
+	Values []*Value
```
[the remark this answers is not in the document]
Comment (MDrakos) [reply pkg/platform/runtime/buildexpression/buildexpression.go:45]: A slice is already a pointer so if the request for this change is so we can do a `nil` check that's already the case. I like the renaming though! I think that's what I originally had but then changed it for some reason 🤔

File: pkg/platform/runtime/buildexpression/buildexpression.go [hunk 2]
```diff
@@ -513,33 +594,82 @@ func getVersionRequirements(v *[]*Value) []model.VersionRequirement {
...
+		return solveAp, nil
+	}
+
+	// Search for solve node in the top level assignments.
+	for _, a := range e.Values {
+		if a.Assignment == nil {
+			continue
+		}
+
+		if a.Assignment.Name == "" && a.Assignment.Name != "runtime" {
```
[the remark this answers is not in the document]
Comment (MDrakos) [reply pkg/platform/runtime/buildexpression/buildexpression.go:614]: Whoops should be an `or`

File: pkg/platform/runtime/buildexpression/buildexpression.go [hunk 3]
```diff
@@ -513,33 +594,82 @@ func getVersionRequirements(v *[]*Value) []model.VersionRequirement {
...
+	for _, a := range let.Assignments {
+		if a.Value == nil {
+			continue
+		}
+
 		if a.Value.Ap == nil {
 			continue
 		}
 
+		if a.Value.Ap.Name == "" && a.Value.Ap.Name != "runtime" {
```
[the remark this answers is not in the document]
Comment (MDrakos) [reply pkg/platform/runtime/buildexpression/buildexpression.go:644]: Same here

File: pkg/platform/runtime/buildexpression/buildexpression.go [hunk 4]
```diff
@@ -715,17 +874,55 @@ func (e *BuildExpression) UpdateTimestamp(timestamp strfmt.DateTime) error {
...
+		if assignment.Value == nil {
+			continue
+		}
+		result[assignment.Name] = assignment.Value
+	}
+
+	result["in"] = l.In
+
+	return nil
+}
```
[the remark this answers is not in the document]
Comment (MDrakos) [reply pkg/platform/runtime/buildexpression/buildexpression.go]: Good point!

File: pkg/platform/runtime/buildexpression/buildexpression.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (mitchell-as) [reply pkg/platform/runtime/buildexpression/buildexpression.go]: I...did not know that. Thanks for making me question an assumption I was making.

**Review by mitchell-as — APPROVED** [review]
Nicely done.

**Review by Naatan — APPROVED** [review]
I see no real architectural changes here, so I'll defer to Mitchell's review.

**Pull request merged**